jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/MainActivity.java

```diff
@@ -1,68 +1,146 @@
+        if (display_message_fragment == null) {
+            FragmentTransaction ft = fragmentManager.beginTransaction();
+            display_message_fragment = new DisplayMessageFragment();
+            ft.add(R.id.fragment_container, display_message_fragment).commitNow();
+        }
+
+    }
+
+    @Override
+    public void onSaveInstanceState(Bundle savedInstanceState) {
+        // Save the state of the DisplayMessageFragment.
+        savedInstanceState.putBoolean(IS_MESSAGE_FRAGMENT_VISIBLE, isDisplayMessageFragmentVisible);
+        super.onSaveInstanceState(savedInstanceState);
     }
 
     /** Called when the user clicks the Send button */
     public void sendMessage(View view) {
 
-        // The message entered by the user.
-        String message = editText.getText().toString();
+        // If the display message fragment is visible show the message there.
+        if (display_message_fragment != null) {
+            display_message_fragment.updateMessage(main_fragment.getMessage());
+        }
 
-        // If orientation is in landscape add message to fragment.
-        if (getResources().getConfiguration().orientation ==
-                Configuration.ORIENTATION_LANDSCAPE) {
+    }
 
-            // Get the fragment that shows the message.
-            if (display_message_fragment == null) {
-                FragmentManager fragmentManager = getSupportFragmentManager();
-                display_message_fragment =
-                        fragmentManager.findFragmentById(R.id.displaymessage_fragment);
-            }
+    /** Called when the user clicks the show/hide button */
+    public void showHide(View view) {
 
-            // Edit the text view (show written message).
-            TextView textView = (TextView)
-                    display_message_fragment.getView().findViewById(R.id.display_message);
-            textView.setText(message);
+        // If the displaying fragment has not been initialized return.
+        if (display_message_fragment == null) {
+            return;
+        }
 
-        // If orientation is in portrait start new activity.
-        } else {
-            // Create intent.
-            Intent intent = new Intent(this, DisplayMessageActivity.class);
+        // Setup FragmentManager and FragmentTransaction.
+        View fragment_container = findViewById(R.id.fragment_container);
+        FragmentManager fragmentManager = getSupportFragmentManager();
+        FragmentTransaction ft = fragmentManager.beginTransaction();
+        ft.setCustomAnimations(android.R.anim.fade_in, android.R.anim.fade_out);
+
+
+        // Hide Fragment if it is visible.
+        if (display_message_fragment.isVisible()) {
+            ft.hide(display_message_fragment).commitNow();
+
+            isDisplayMessageFragmentVisible = false;
+
+            // Inform the main fragment of the change
+            main_fragment.setButtonText();
```

Comment:
Is there a nice way to create a listener on the `isDisplayMessageFragmentVisible` so that I don't have to call `setButtonText()` when I change the value? They show a way [here](https://developer.android.com/training/basics/fragments/communicating.html), but it seems a bit overkill here.